sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -136,9 +132,11 @@ internal static IEnumerable<BatchedLogs> Batch<T>(IEnumerable<T> logEntries, Log
                     tempWriter.WriteEndArray();
                     tempWriter.Flush();
                     yield return new BatchedLogs(new List<object> { log }, BinaryData.FromBytes(tempArrayBuffer.WrittenMemory));
+                    continue;
                 }
-                // if adding this entry makes stream > 1 Mb send current stream now
-                else if ((writer.BytesPending + memory.Length + 1) >= SingleUploadThreshold)
+
+                // if adding this entry (and array end) would make stream > 1 Mb send current stream now
+                if ((writer.BytesCommitted + writer.BytesPending + memory.Length + 2) > SingleUploadThreshold)
```

Comment:
@Arithmomaniac Why are we including BytesCommitted in this check? BytesCommitted represents anything that we flushed and we always yield return the batched logs right after flushing so BytesCommitted is always 0. BytesPending holds the current number of bytes we haven't flushed yet.

Reply:
`Uft8JsonWriter` reserves the right to flush without user intervention. In fact, it will do this if its internal buffer is ever too small to hold a value to be serialized:

https://github.com/dotnet/runtime/blob/ec4437be46d8b90bc9fa6740c556bd860d9fe5ab/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs#L1084-L1086

I don't remember exactly how I stumbled across this during testing (maybe while building the million-char-long strings used in the unit tests), but in any case it's not safe to assume `BytesCommitted` is always zero.